src/util/transaction.rs

```diff
@@ -114,3 +114,237 @@ where
     s.serialize_field("vout", &outpoint.vout)?;
     s.end()
 }
+
+pub(super) mod sigops {
+    use crate::chain::{
+        hashes::hex::FromHex,
+        opcodes::{
+            all::{OP_CHECKMULTISIG, OP_CHECKMULTISIGVERIFY, OP_CHECKSIG, OP_CHECKSIGVERIFY},
+            All,
+        },
+        script::{self, Instruction},
+        Transaction, TxOut, Witness,
+    };
+    use std::collections::HashMap;
+
+    /// Get sigop count for transaction. prevout_map must have all the prevouts.
+    pub fn transaction_sigop_count(
+        tx: &Transaction,
+        prevout_map: &HashMap<u32, &TxOut>,
+    ) -> Result<usize, script::Error> {
+        let input_count = tx.input.len();
+        let mut prevouts = Vec::with_capacity(input_count);
+        for idx in 0..input_count {
+            prevouts.push(
+                *prevout_map
+                    .get(&(idx as u32))
+                    .ok_or(script::Error::EarlyEndOfScript)?,
+            );
+        }
+
+        get_sigop_cost(tx, &prevouts, true, true)
+    }
+
+    fn decode_pushnum(op: &All) -> Option<u8> {
+        // 81 = OP_1, 96 = OP_16
+        // 81 -> 1, so... 81 - 80 -> 1
+        let self_u8 = op.into_u8();
+        match self_u8 {
+            81..=96 => Some(self_u8 - 80),
+            _ => None,
+        }
+    }
+
+    fn count_sigops(script: &script::Script, accurate: bool) -> Result<usize, script::Error> {
+        let mut n = 0;
+        let mut pushnum_cache = None;
+        for inst in script.instructions() {
+            match inst? {
+                Instruction::Op(opcode) => {
+                    match opcode {
+                        OP_CHECKSIG | OP_CHECKSIGVERIFY => {
+                            n += 1;
+                        }
+                        OP_CHECKMULTISIG | OP_CHECKMULTISIGVERIFY => {
+                            match (accurate, pushnum_cache) {
+                                (true, Some(pushnum)) => {
+                                    // Add the number of pubkeys in the multisig as sigop count
+                                    n += usize::from(pushnum);
+                                }
+                                _ => {
+                                    // MAX_PUBKEYS_PER_MULTISIG from Bitcoin Core
+                                    // https://github.com/bitcoin/bitcoin/blob/v25.0/src/script/script.h#L29-L30
+                                    n += 20;
+                                }
+                            }
+                        }
+                        _ => {
+                            pushnum_cache = decode_pushnum(&opcode);
```

Comment:
is it worth deferring the `decode_pushnum` call until we match an `OP_CHECKMULTISIG`, the same way Core does it?

Reply:
Not really... this will never be a performance bottleneck. lol